src/index.ts

```diff
@@ -1,173 +1,231 @@
-    if (shouldShowTooltip(item) && tooltipActive) {
-      update(event, item, options);
+    if (title) {
+      content += `<h2>${title}</h2>`;
     }
-  });
-
-  // clear tooltip on mouse out
-  vgView.addEventListener('mouseout.tooltipRemove', function (event: MouseEvent, item: Scenegraph) {
-    if (shouldShowTooltip(item)) {
-      cancelPromise();
 
-      if (tooltipActive) {
-        clear(event, item, options);
+    content += '<table>';
+    for (const key of Object.keys(rest)) {
+      let val = (rest as any)[key];
+      if (isObject(val)) {
+        val = stringify(val);
       }
+
+      content += `<tr><td class="key">${key}:</td><td class="value">${val}</td></tr>`;
     }
-  });
-}
+    content += `</table>`;
+
+    return content;
+  }
 
-/* Cancel tooltip promise */
-function cancelPromise() {
-  /* We don't check if tooltipPromise is valid because passing
-   an invalid ID to clearTimeout does not have any effect
-   (and doesn't throw an exception). */
-  window.clearTimeout(tooltipPromise);
-  tooltipPromise = undefined;
+  return `${value}`;
 }
 
-/* Initialize tooltip with data */
-function init(event: MouseEvent, item: Scenegraph, options: Option): void {
-  // get tooltip HTML placeholder
-  const tooltipPlaceholder = getTooltipPlaceholder();
+/**
+ * Attach a tooltip handler to the view.
+ */
+export default function(view: View, opt?: Options) {
+  let visible = false;
+
+  const options = { ...DEFAULT_OPTIONS, ...opt };
 
-  // prepare data for tooltip
-  const tooltipData = getTooltipData(item, options);
-  if (!tooltipData || tooltipData.length === 0) {
-    return undefined;
+  init(options);
+
+  function hideTooltip() {
+    tooltipElement.classList.remove('visible', `${options.theme}-theme`);
+    visible = false;
   }
 
-  // bind data to tooltip HTML placeholder
-  bindData(tooltipPlaceholder, tooltipData);
+  function tooltipHandler(this: View, handler: any, event: MouseEvent, item: any, value: any) {
+    // console.log(this, handler, event, item, value);
 
-  updatePosition(event, options);
-  updateColorTheme(options);
-  select('#vis-tooltip').style('visibility', 'visible');
-  tooltipActive = true;
+    if ((event as any).vegaType === undefined) {
+      hideTooltip();
+      return;
+    }
 
-  // invoke user-provided callback
-  if (options.onAppear) {
-    options.onAppear(event, item);
-  }
-}
+    tooltipElement.innerHTML = formatValue(value);
 
-/* Update tooltip position on mousemove */
-function update(event: MouseEvent, item: Scenegraph, options: Option): void {
-  if (!shouldShowTooltip(item)) {
-    return undefined;
+    tooltipElement.classList.add('visible', `${options.theme}-theme`);
+    visible = true;
   }
-  updatePosition(event, options);
 
-  // invoke user-provided callback
-  if (options.onMove) {
-    options.onMove(event, item);
-  }
-}
+  function handleMouseMove(event: ScenegraphEvent, item?: Item) {
```

Comment:
This makes sense for the time being, however, you want (or need) to update this when the next Vega release comes out. In order to support tooltips over area/line/trail marks, I needed to make tooltips trigger on `mousemove` (previously only `mouseover`/`mouseout` were observed).

You'll still want to perform x/y layout, but ultimately you may only need to do this from the tooltip handler, without need for any additional event listeners!

Reply:
I'll keep this in mind. It will definitely make the code easier.

Reply:
Is there a way for tooltip handler to detect if this is the same object or a different one?  E.g. should the tooltip show up and not move/disappear until the user moves to a different Vega mark, or should the tooltip always move?   If tooltip stays in one place, it may be possible to move the mouse over the tooltip and copy some of that data.  Might be an optional behavior.

Reply:
Oh I hate how twitter handles this. 

![twitter](https://user-images.githubusercontent.com/589034/39090100-ed681716-4589-11e8-92d9-9a3a5dcda013.gif)